### PR TITLE
Models

### PR DESCRIPTION
**Pull request by anakib1**

### CODE REVIEW

File: mangoPURE/models/base.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import torch
+from ..data.base import TimedAudioBatch, MixedExample
+from typing import Any
+from dataclasses import dataclass
+
+
+class Embedder(torch.nn.Module):
+    def forward(self, batch: TimedAudioBatch) -> TimedAudioBatch:
```
Comment (anakib1) [inline mangoPURE/models/base.py]: Why does it return TimedAudioBatch and not torch.Tensor?

File: mangoPURE/models/collators.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    def __call__(self, batch_list: list[MixedExample]) -> TimedAudioBatch:
+        audio_list = list(map(lambda x: x.audio, batch_list))
+        lengths = [x.shape[0] for x in audio_list]
+        start_size = lengths[0]
+        noises_info = None if batch_list[0].noises_info is None \
+            else list(map(lambda x: x.noises_info, batch_list))
+        speakers_info = None if batch_list[0].speakers_info is None \
+            else list(map(lambda x: x.speakers_info, batch_list))
+        audio_features = self.feature_extractor(list(map(lambda x: x.numpy(), audio_list)),
```
Comment (anakib1) [inline mangoPURE/models/collators.py]: We don't need to map torch.tensors to numpy to map them to torch.tensor back. Let's use torch.tensors everywhere (in resample you also used librosa that is also bad, better use torch everywhere, this part of code (generation) is the slowest one in pipeline

File: mangoPURE/models/collators.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                                                return_tensors="pt", return_attention_mask=True, sampling_rate=RATE)
+        audio_features_batch = self.feature_extractor.pad(audio_features)
+        end_size = audio_features_batch["attention_mask"][0].sum().tolist()
```
Comment (anakib1) [inline mangoPURE/models/collators.py]: And why is it final size? Should not we pad to some constant length and truncate?

File: mangoPURE/models/metrics.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+class SigmoidTimedLoss(TimedLoss):
+    def __init__(self, info_type: str = "noise"):
+        """
+        :param info_type: ["noise", "speaker"] selects which metadata from batch to use
+        """
+        self.info_type = info_type
+
+    def __call__(self, batch: TimedAudioBatch) -> torch.Tensor:
+        mask = torch.zeros(batch.audio.shape)
+        for i, length in enumerate(batch.lengths):
```
Comment (anakib1) [inline mangoPURE/models/metrics.py]: for is slow af , better write alchemic operation mask[magic(batch.lengths)] (ask got)

File: mangoPURE/models/metrics.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            mask[i, :length, :] = 1.0
+        labels = torch.zeros(batch.audio.shape)
```
Comment (anakib1) [inline mangoPURE/models/metrics.py]: zeros_like 😸

File: mangoPURE/models/metrics.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        labels[:, :, 0] = 1.0
+        batch_segments = None
+        if self.info_type == "noise":
+            batch_segments = batch.noises_info
+        elif self.info_type == "speaker":
+            batch_segments = batch.speakers_info
```
Comment (anakib1) [inline mangoPURE/models/metrics.py]: Both are not supported?

File: mangoPURE/models/metrics.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        assert batch_segments is not None, "segments is None for some reason"
+        for i, segments in enumerate(batch_segments):
+            for segment in segments:
+                labels[i, segment.start:segment.start + segment.length, 0] = 0.0
+                labels[i, segment.start:segment.start + segment.length, segment.class_id] = 1.0
```
Comment (anakib1) [inline mangoPURE/models/metrics.py]: This should be generated in collator. This is bad when models does something like that in training, it slows down everything and we can't do anything about it. Labels should be already prepared on this state

File: mangoPURE/models/modules.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        batch.audio = self.model(batch.audio).last_hidden_state
+        new_size = batch.audio.shape[1]
+        batch = Resize.resize_timed_audio_batch(batch, start_size, new_size)
+        return batch
+
+
+class LinearTimedHead(TimedHead):
+    def __init__(self, input_dim, num_classes):
+        super().__init__(input_dim, num_classes)
+        self.model = torch.nn.Linear(input_dim, num_classes)
```
Comment (anakib1) [inline mangoPURE/models/modules.py:26]: Maybe activation and then linear?

Comment (anakib1) [reply mangoPURE/models/modules.py:26]: Nvm, seems like last output of transformer is after activation

File: mangoPURE/models/base.py (hunk 1 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/base.py]: Because the embedder should pass the information about the noise/speaker segments

File: mangoPURE/models/collators.py (hunk 2 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/collators.py]: At first I tried to do it, unfortunately (if I am not mistaken, and haven't made stupid mistake in the code) WhisperFeatureExtractor does not support torch.Tensor in the __call__ function. MixedExample requires to audio in torch.Tensor format, that is why I haven't found appropriate solution here. What do you suggest to do?
Look <a href="https://huggingface.co/docs/transformers/v4.39.2/en/model_doc/whisper#transformers.WhisperFeatureExtractor">here</a> for __call__ method docs

File: mangoPURE/models/collators.py (hunk 3 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/collators.py]: I have checked, I made a mistake and 23 line is just useless altogether (I just need to delete it, because it doesn't do anything)
I don't understand the question, but maybe this is the answer: Whisper encoder (despite the fact that it is a part of transformer) requires to have an input of constant length. Also, as of my knowledge, whisper extractor pads the data itself

File: mangoPURE/models/metrics.py (hunk 4 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/metrics.py]: Reasonable note. I will search for a solution

File: mangoPURE/models/metrics.py (hunk 5 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/metrics.py]: Thank you, I haven't known this

File: mangoPURE/models/metrics.py (hunk 6 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/metrics.py]: Heh, I haven't thought about that. But no, I think I have to delete this **info_type** field and just check if **noises_info** or **speakers_info** 8 are present in batch, then any combination will be supported (except the combination, where both are None, of course I will raise an Exception then)

File: mangoPURE/models/metrics.py (hunk 7 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/metrics.py]: You have a point, but I don't know how to implement it, because the shape of labels depends on the model. If you want me to implement this in collator, than the collator will depend on the model.

File: mangoPURE/models/collators.py (hunk 3 above)
Comment (anakib1) [reply mangoPURE/models/collators.py]: why end size is number of ones in the first attention mask?

File: mangoPURE/models/utils.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+class Whisper:
+
+    @staticmethod
+    def update_segment_infos(audio: torch.Tensor, infos: List[SegmentInfo]):
+        """
+        Transforms segment infos according to whisper feature extractor output
+        !!!updates info in-place
+        """
+        num_secs = audio.shape[0] / RATE
+        if num_secs > 30.0:
```
Comment (anakib1) [inline mangoPURE/models/utils.py]: Bad

Comment (anakib1) [reply mangoPURE/models/utils.py]: better pass this value somewhere and not hardcode

File: mangoPURE/models/modules.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        super().__init__()
+        self.model = WhisperModel.from_pretrained(hf_name).encoder
+
+    def forward(self, batch: Dict) -> Dict:
+        """
+        :param batch: dictionary that should have key "input_features"
+            that whisper encoder takes
+        :return: batch with whisper embedder output ("embeddings" ket in batch)
+        """
+        batch["embeddings"] = self.model(input_features=batch["input_features"]).last_hidden_state
```
Comment (anakib1) [inline mangoPURE/models/modules.py:19]: 🤓 better use penultimate layer

File: mangoPURE/models/modules.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+        del batch["input_features"]
```
Comment (anakib1) [inline mangoPURE/models/modules.py]: why? It can break other modules that use the same batch. Model should not break it

File: mangoPURE/models/modules.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        self.model = torch.nn.Linear(input_dim, num_classes)
+
+    def forward(self, batch: Dict) -> Dict:
+        """
+        :param batch: dictionary that should have key "embeddings",
+            embedder module usually provide this data
+        :return: batch with head output ("head_output" key in batch)
+        """
+        batch["head_output"] = self.model(batch["embeddings"])
+        del batch["embeddings"]
```
Comment (anakib1) [inline mangoPURE/models/modules.py]: same here

File: mangoPURE/models/metrics.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+        """
+        :param batch: dictionary with params:
+            "attention_mask" - the mask of audio embeddings
+            "labels" - true labels of the model
+            "head_output" - last output of the model
+        :return: binary cross entropy loss
+        """
+        mask = batch["attention_mask"]
+        loss_sum = mask * torch.nn.BCEWithLogitsLoss()(batch["head_output"], batch["labels"])
+        loss = loss_sum.sum() / mask.sum()
```
Comment (anakib1) [inline mangoPURE/models/metrics.py]: image

Does it work? Seems like BCEWithLogitsLoss returns 0-dim tensor

File: mangoPURE/models/utils.py (hunk 9 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/utils.py]: okay

File: mangoPURE/models/metrics.py (hunk 13 above)
Comment (ZaraGiraffe) [reply mangoPURE/models/metrics.py]: Omg, thank you, we were very close to a disaster
